sycl/doc/extensions/experimental/sycl_ext_oneapi_copy_optimize.asciidoc

```diff
@@ -0,0 +1,139 @@
+
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback. Interfaces defined in this
+specification are implemented in DPC\++ but they are not finalized and may
+change incompatibly in future versions of DPC++ without prior notice.
+Shipping software products should not rely on APIs defined in this
+specification.
+
+
+== Overview
+
+SYCL provides explicit copy APIs that copy memory between a device and the host.
+Some backends can optimize these copy operations if the backend knows in
+advance that the host memory will be used for such a copy operation.
+If a SYCL application knows that it will frequently copy to or from the same
+host memory, it can use the APIs in this extension to inform the backend
+about this, which can result in a performance benefit.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine
+the macro `SYCL_EXT_ONEAPI_USM_IMPORT_RELEASE` to one of the values defined
+in the table below.  Applications can test for the existence of this macro
+to determine if the implementation supports this feature, or applications
+can test the macro's value to determine which of the extension's features
+the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API of the extension
+
+This extension adds the following free functions:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+void prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                             const context &Context);
+void prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                             const queue &Queue);
+
+void release_from_device_copy(void *Ptr, const context &Context);
+void release_from_device_copy(void *Ptr, const queue &Queue);
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+Table 1. Functions added by this extension.
+|===
+| Member Function | Description
+
+| `prepare_for_device_copy(void *Ptr, size_t NumBytes, const context &Context)`.
+| Informs the implementation that the host memory range starting at `Ptr` and
+extending for `NumBytes` bytes may be used as either the source or destination
+of an explicit copy operation to a queue using context Context. Calling this
+function may allow the implementation to accelerate these copy operations.
+Generally, it is only useful to call this function if the host memory range will
+be used for several subsequent copy operations; it is not generally useful to
+call it before each individual copy operation.
+
+Although the context is specified in the API the current implementation does the
+import in the GPU Driver where SYCL context is unavailable.
+Therefore, it is valid to call `prepare_for_device_copy` on the same host memory
+from different SYCL contexts. 
```

Comment:
The first sentence about the current implementation not using the context is an implementation detail.  The spec should define the API that you are committed to support even for future implementations.

I assume you must be somehow using the context, right?  How else do you know which backend is being used?  For example, you wouldn't want to call `zexDriverImportExternalPointer` if the context is for the CUDA backend.

The sentence above is also contradictory to the sentence below.  The above sentence says it is OK to call `prepare_for_device_copy` for the same memory range with two different contexts, but the sentence below says that two calls may not have overlapping memory ranges.  Passing the same range to two calls violates the second sentence which says that ranges can't overlap.

I'd suggest combining these last two paragraphs into a new paragraph with some clearer statement about overlapping memory ranges and contexts.  Options I see are:

* Do not allow overlapping memory ranges at all, even for different context or backends.
* Allow overlapping memory ranges only when the context backend is different.
* Allow overlapping memory ranges only when the context is different.
* Allow overlapping memory ranges regardless of the context.

Reply:
All suggested doc changes have been made, and the macro name has been changed also.